CSS3 transitions for all fade animations

apenchev opened the pull request:
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #37691.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Add this custom code to your theme's `functions.php` in order to enable fade animations:
```
function carousel_options( $options ) {
	return array_merge(
		$options,
		array(
			'animation' => 'fade',
			'smoothHeight' => false, // smoothHeight is extremely buggy with fade for some reason
			'animationSpeed' => 2600,
			'fadeFirstSlide' => false,
		)
	);
}
add_filter( 'woocommerce_single_product_carousel_options', 'carousel_options' );
```
2. Visit a Single Product Page on a non-touch-enabled device (a PC or a laptop).
3. Confirm that the first gallery image does NOT fadeIn on page load.
4. Confirm that switching between gallery images happens with a fade animation that lasts 2.6 seconds.
5. Enable `fadeFirstSlide` in the custom code in your theme's `functions.php`:
```
function carousel_options( $options ) {
	return array_merge(
		$options,
		array(
			'animation' => 'fade',
			'smoothHeight' => false, // smoothHeight is extremely buggy with fade for some reason
			'animationSpeed' => 2600,
			'fadeFirstSlide' => true,
		)
	);
}
add_filter( 'woocommerce_single_product_carousel_options', 'carousel_options' );
```
6. Visit a Single Product Page on a non-touch-enabled device (a PC or a laptop).
7. Confirm that the first gallery image DOES fadeIn on page load.

`'smoothHeight' => false,` is needed since `smoothHeight` doesn't play nice with `fade` - that's an existing bug with FlexSlider.

<!-- End testing instructions -->

apenchev committed:
CSS3 transitions for all fade animations

apenchev committed:
Changelog

apenchev committed:
Whitespace

apenchev committed:
Merge branch 'woocommerce:trunk' into flexcss

apenchev commented:
@vedanshujain Would you be able to take a look?

apenchev committed:
Formatting

apenchev committed:
Merge branch 'flexcss' of https://github.com/apenchev/woocommerce into flexcss

apenchev committed:
Lambda function

apenchev committed:
Comment

apenchev committed:
Merge branch 'woocommerce:trunk' into flexcss

apenchev commented:
:/

apenchev committed:
Merge branch 'woocommerce:trunk' into flexcss

apenchev commented:
> @vedanshujain Would you be able to take a look?

Konamiman commented:
Testing works as expected, but I'm requesting analysis since I'm not really expert in JavaScript, thanks.

alopezari commented:
Thanks @Konamiman!

Will make sure to test this thoroughly once it gets included in the next release package 👍

gigitux commented on plugins/woocommerce/client/legacy/js/flexslider/jquery.flexslider.js:
```diff
@@ -720,12 +720,14 @@
-            slider.slides.eq(slider.currentSlide).css({"zIndex": 1}).animate({"opacity": 0}, slider.vars.animationSpeed, slider.vars.easing);
-            slider.slides.eq(target).css({"zIndex": 2}).animate({"opacity": 1}, slider.vars.animationSpeed, slider.vars.easing, slider.wrapup);
+            // Unbind previous transitionEnd events and re-bind new transitionEnd event
+            slider.slides.eq(slider.currentSlide).off("transitionend");
```
Is there a reason for this change? Could it be an issue for "extensions" that add a handler on the transitionend?

apenchev replied:
Thanks for testing! Old code used to call `jQuery.animate()` and pass `slider.wrapup` as [the `complete` callback](https://api.jquery.com/animate/#animate-properties-duration-easing-complete-complete). In order to replicate [the `complete` callback from jQuery](https://api.jquery.com/animate/#animate-properties-duration-easing-complete-complete), we now call `slider.wrapup` during the CSS `transitionend`.

If there are plugins subscribing to `transitionend` on `slider.slides`, they are doing it knowing that the subscription would only work on touch-enabled devices. Remember, in the production code CSS transitions are ONLY used inside `if (touch) { }`.

In this PR, we remove all current `transitionend` subscriptions with `.off("transitionend")` inside `if (!touch) { }`.

gigitux replied:
Hi, @apenchev - thanks for the explanation. Just to be sure. Given this:

> If there are existing plugins subscribing to transitionend on slider.slides, they are doing it knowing that the subscription would only work on touch-enabled devices (remember, before this PR, CSS transitions are ONLY used inside if (touch) { })

> inside of if (!touch) { } we remove existing transitionend subscriptions before adding our own transitionend subscription

It means that this PR doesn't introduce any potential breaking change. Right?

apenchev replied:
Unfortunately, the PR does introduce a potential breaking change. My explaination was not correct, as I assumed that we do a .transitionend cleanup inside of `if (!touch) { }`, but we don't clean up everything.

I will try and send a small fix that attempts to do a full .transitionend cleanup inside of `if (!touch) { }`.

apenchev replied:
I've pushed the commit with the proper transitionend cleanup.

gigitux reviewed (commented):
Thanks for working on this! 

I've thoroughly tested the pull request and it looks good to me. However, my primary concern is to ensure that this PR won't cause any regressions or disrupt existing plugins or themes. I've pinpointed a specific concern with line 728, and I'd like to hear your insights before granting my approval.

apenchev committed:
Merge branch 'woocommerce:trunk' into flexcss

apenchev committed:
Clean up transitionEnd event listeners if (!touch)

apenchev committed:
Update this.offsetWidth comment

gigitux reviewed (changes requested):
Hi @apenchev,
Sorry for the late answer!
Thanks for your work in addressing the feedback and improving the code to avoid any potential breaking changes.

I noticed that with the latest changes, the thumbnails aren't visible anymore:


https://github.com/woocommerce/woocommerce/assets/4463174/f399e673-6238-4b5f-8df6-b52b2d81e98a

Can you check and see if you are able to replicate it?

apenchev commented:
> 'animationSpeed' => 1600,

I can't reproduce it, looks good on my end. :/ I'm using the StoreFront default theme, WooCommerce status is:

WooCommerce: 8.3.1
WooCommerce REST API: 8.3.1
WooCommerce Blocks package: 11.4.9
Action Scheduler package: 3.6.4
WordPress: 6.4.1

apenchev commented:
> Hi @apenchev, Sorry for the late answer! Thanks for your work in addressing the feedback and improving the code to avoid any potential breaking changes.
> 
> I noticed that with the latest changes, the thumbnails aren't visible anymore:
> 
>  Screen.Capture.Nov.14.mov 
> Can you check and see if you are able to replicate it?

Yes, the bug happens on the Twenty Twenty-Four theme when you set the animation to 'fade' whether you're using this PR or not. ;D I'm using the StoreFront theme for testing so I didn't see this bug.

In conculustion, this PR doesn't add this bug, I seem to get it even on a clean install simply by adding:


```
function carousel_options( $options ) {
	return array_merge(
		$options,
		array( 'animation' => 'fade' )
	);
}
add_filter( 'woocommerce_single_product_carousel_options', 'carousel_options' ); 
```

apenchev committed:
Use jQuery methods in order to greatly reduce diff

gigitux reviewed (commented):
Thanks for your effort on this PR! It is LGTM and it is ready to be shipped. Before approving the PR, can you rebase it with trunk?

apenchev committed:
Merge branch 'woocommerce:trunk' into flexcss

apenchev commented:
trunk merged.

gigitux reviewed (approved):
Thanks again for your effort! I'll merge the PR once all CI checks pass.
